source/slang/slang-ir-extract-value-from-type.cpp

```diff
@@ -0,0 +1,278 @@
+        break;
+    default:
+        {
+            result.leafValue = src;
+            result.offsetInValue = offset;
+            result.valueSize = (uint32_t)layout.size;
+            return result;
+        }
+        break;
+    }
+}
+
+IRInst* extractByteAtOffset(
+    IRBuilder& builder,
+    TargetRequest* targetReq,
+    IRType* dataType,
+    IRSizeAndAlignment& layout,
+    IRInst* src,
+    uint32_t offset)
+{
+    auto leaf = findLeafValueAtOffset(targetReq, builder, dataType, layout, src, offset);
+    IRType* uintType = nullptr;
+    if (leaf.valueSize <= 4)
+    {
+        uintType = builder.getUIntType();
+    }
+    else
+    {
+        uintType = builder.getUInt64Type();
+    }
+    auto resultValue = builder.emitBitCast(uintType, leaf.leafValue);
+    if (leaf.offsetInValue != 0)
+    {
+        uint32_t shift = leaf.offsetInValue * 8;
+        resultValue = builder.emitShr(uintType, resultValue, builder.getIntValue(uintType, shift));
+
+        resultValue = builder.emitBitAnd(
+            builder.getUIntType(),
+            resultValue, builder.getIntValue(builder.getUIntType(), 0xFF));
+    }
+    return resultValue;
+}
+
+IRInst* extractMultiByteValueAtOffset(
+    IRBuilder& builder,
+    TargetRequest* targetReq,
+    IRType* dataType,
+    IRSizeAndAlignment& layout,
+    IRInst* src,
+    uint32_t size,
+    uint32_t offset)
+{
+    if (size == 1)
+        return extractByteAtOffset(builder, targetReq, dataType, layout, src, offset);
+
+    auto leaf = findLeafValueAtOffset(targetReq, builder, dataType, layout, src, offset);
+    auto resultValue = leaf.leafValue;
+    IRType* uintType = nullptr;
+    if (leaf.valueSize <= 4)
+    {
+        uintType = builder.getUIntType();
+    }
+    else
+    {
+        uintType = builder.getUInt64Type();
+    }
+    if (leaf.valueSize - leaf.offsetInValue >= size)
+    {
+        // The request value is fully contained in the found leaf element.
+        // We can proceed to extract the requested bits from the element.
+        uint32_t shift = leaf.offsetInValue * 8;
+        if (shift > 0)
+            resultValue = builder.emitShr(uintType, resultValue, builder.getIntValue(uintType, shift));
+        uint32_t bitMask = 0;
+        switch (size)
+        {
+        case 1:
+            bitMask = 0xFF;
+            break;
+        case 2:
+            bitMask = 0xFFFFF;
+            break;
+        case 3:
+            bitMask = 0xFFFFFF;
+            break;
+        case 4:
+            bitMask = 0xFFFFFFFF;
+            break;
+        default:
+            break;
+        }
+        if (leaf.valueSize != size)
+        {
+            resultValue =
+                builder.emitBitAnd(uintType, resultValue, builder.getIntValue(uintType, bitMask));
+        }
+        return resultValue;
+    }
+    else
+    {
+        // The requested value crosses the boundaries of different fields.
+        // We need to extract first and second half separately, and combine them together.
```

Comment:
In principle, we might have to extract more than twice, right?

It seems like the logic here eventually needs to be something like:

```
remainingBits = totalSizeInBits;
foundBits = 0;
offset = ...;
accumulator = nullptr;
while(remainingBits) {
    leafValue = findleafValue(offset);
    leafBits = leafValue.bits;
    if(leafBits > remainingBits) {
        leafBits = remainingBits;
        leafValue = mask(leafValue, leafBits);
    }
    if(foundBits > 0) {
        leafValue = shiftLeft(leafValue, foundBits);
    }
    accumulator = accumulator == null ? leafValue : or(leafValue, accumulator);
    foundBits += leafBits;
    remainingBits -= leafBits;
}
return accumulator
```

Reply:
The call itself can recurve more than once to take care of all the cases.

Reply:
Ah, I hadn't thought about that. So long as all the leaf values are power-of-two sizes, then that approach would seemingly ensure a minimal number of reads. Nice.